lib/src/main/java/com/knuddels/jtokkit/TokenEncoder.java

```diff
@@ -1,80 +1,196 @@
-        for (Map.Entry<T, Integer> entry : input.entrySet()) {
-            ImmutableByteArray key = keyMapper.apply(entry.getKey());
-            Integer value = entry.getValue();
-            decodedToEncoded.put(key, value);
+    public int addTokensAndGetCount(Integer maxTokenCount, byte[] utf8Bytes, List<Integer> out, List<Integer> ranks) {
+        ImmutableByteArray match = ImmutableByteArray.from(utf8Bytes);
+        int encoded = encode(match);
+        if (encoded != MAX_RANK) {
+            out.add(encoded);
+            return 1;
+        } else {
+            int length = match.length();
+            return addTokensAndGetCount(maxTokenCount, out, ranks, match, length);
         }
     }
 
-    /**
-     * Checks if the given decoded token is contained in this encoder.
-     *
-     * @param decodedToken the decoded token
-     * @return true if the decoded token is contained in this encoder, false otherwise
-     */
-    public boolean containsDecodedToken(ImmutableByteArray decodedToken) {
-        return decodedToEncoded.containsKey(decodedToken);
+    private int addTokensAndGetCount(Integer maxTokenCount, List<Integer> out, List<Integer> ranks, ImmutableByteArray match, int length) {
+        int validRanks = initRanks(match, length, ranks);
+        int tokenCount = mergeBytesAndGetTokenCount(match, length, ranks, validRanks);
+        for (int start = 0, end = 1; end < ranks.size() && (maxTokenCount == null || out.size() < maxTokenCount); end++) {
+            if (ranks.get(end) != DUMMY_RANK) {
+                int token = encode(match, start, end);
+                assert token != MAX_RANK : "Token should not be MAX_RANK";
+                out.add(token);
+                start = end;
+            }
+        }
+        return tokenCount;
     }
 
-    /**
-     * Encodes the given decoded token.
-     *
-     * @param decodedToken the decoded token
-     * @return the encoded token
-     * @throws IllegalArgumentException if the decoded token is not contained in this encoder
-     */
-    public Integer encode(ImmutableByteArray decodedToken) {
-        Integer encoded = decodedToEncoded.get(decodedToken);
-        if (encoded == null) {
-            throw new IllegalArgumentException("Unknown token for encoding: " + decodedToken);
+    int initRanks(ImmutableByteArray piece, int tokenCount, List<Integer> ranks) {
+        assert tokenCount > 1 : "Already filtered out";
+        ranks.clear();
+        int validRanks = 0;
+        for (int i = 0; i < tokenCount + 1; i++) {
+            int encoded = encode(piece, i, i + 2);
+            if (encoded != MAX_RANK) {
+                validRanks++;
+            }
+            ranks.add(encoded);
         }
+        return validRanks;
+    }
+
+    int mergeBytesAndGetTokenCount(ImmutableByteArray piece, int length, List<Integer> ranks, int validRanks) {
+        assert true;
+        while (true) {
+            if (validRanks == 0) {
```

Comment:
Iteration stops before the last minimum is MAX_RANK by keeping track of merge results - resulting in one less minimum search at the end